WordPress/Classes/Services/MediaImportService.swift

```diff
@@ -20,55 +20,254 @@ open class MediaImportService: LocalCoreDataService {
+    ///         object, on failure it will return a nil Media and an error object with the details.
+    ///
+    /// - Returns: A NSProgress that tracks the progress of the export process
+    ///
+    /// - SeeAlso: `createMedia(with:blog:post:thumbnailCallback:completion:)`
+    func createMedia(
+        with exportable: ExportableAsset,
+        blog: Blog,
+        post: AbstractPost?,
+        thumbnailCallback: ((Media, URL) -> Void)?,
+        completion: @escaping (Media?, Error?) -> Void
+    ) -> (Media, Progress)? {
+        assert(Thread.isMainThread, "\(#function) can only be called from the main thread")
+
+        guard let media = try? createMedia(with: exportable, blogObjectID: blog.objectID, postObjectID: post?.objectID, in: coreDataStack.mainContext) else {
+            return nil
+        }
+
+        coreDataStack.saveContextAndWait(coreDataStack.mainContext)
+
+        let blogInContext: Blog
+        do {
+            blogInContext = try coreDataStack.mainContext.existingObject(with: blog.objectID) as! Blog
+        } catch {
+            completion(nil, error)
+            return nil
+        }
+
+        let createProgress = self.import(exportable, to: media, blog: blogInContext, thumbnailCallback: thumbnailCallback) {
+            switch $0 {
+            case let .success(media):
+                completion(media, nil)
+            case let .failure(error):
+                completion(nil, error)
+            }
+        }
+
+        return (media, createProgress)
+    }
+
+    /// Create a media object using the url provided as the source of media.
+    ///
+    /// Unlike `createMedia(with:blog:post:thumbnailCallback:completion:)`, this function can be called from any thread.
+    ///
+    /// - Note: All blocks arguments are called from the main thread. The `Media` argument in the blocks is bound to
+    ///     the main context.
+    ///
+    /// - Parameters:
+    ///   - exportable: an object that implements the exportable interface
+    ///   - blog: the blog object to associate to the media
+    ///   - post: the post object to associate to the media
+    ///   - progress: a NSProgress that tracks the progress of the export process.
+    ///   - receiveUpdate: a block that will be invoked with the created `Media` instance.
+    ///   - thumbnailCallback: a block that will be invoked when the thumbail for the media object is ready
+    ///   - completion: a block that will be invoked when the media is created, on success it will return a valid Media
+    ///         object, on failure it will return a nil Media and an error object with the details.
+    @objc
+    @discardableResult
+    func createMedia(
+        with exportable: ExportableAsset,
+        blog: Blog,
+        post: AbstractPost?,
+        receiveUpdate: ((Media) -> Void)?,
+        thumbnailCallback: ((Media, URL) -> Void)?,
+        completion: @escaping (Media?, Error?) -> Void
```

Comment:
The code never has both `Media` and `Error` `nil` at the same time. However, I'm guessing this doesn't use `Result` because of `@objc`?